src/backend/optimizer/plan/createplan.c

```diff
@@ -4456,9 +4463,12 @@ create_ctescan_plan(PlannerInfo *root, Path *best_path,
 			 * ShareInputScan as writing slice creator, in order to prevent
 			 * the situation, when consumer gets to the writer gang and producer
 			 * gets to the reader gang (it depends of tree traverse order
-			 * inside the apply_shareinput_dag_to_tree function)
+			 * inside the apply_shareinput_dag_to_tree function). This only
+			 * applies on QD, where slice and gangType are assigned.
 			 */
-			if (root->curSlice->gangType != saved_gangType &&
+			if (Gp_role == GP_ROLE_DISPATCH &&
+				root->curSlice != NULL &&
```

Comment:
The same thing, if we assume that `curSlice` is not NULL with the dispatch role, it makes no sense to check both.

Reply:
See my previous answer.